File: Languages/go/algorithms-dsa/leetcode/linked-list/linked_list_cycle.go
/*
LeetCode #141: Linked List Cycle
Topic: Linked List
Difficulty: Easy

Given head, determine if the linked list has a cycle in it.

Approach: Floyd's cycle detection - slow pointer moves one step, fast
pointer moves two steps. If they meet, there's a cycle.

Time: O(n), Space: O(1)
*/
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return true
		}
	}
	return false
}

func main() {
	// Test 1: cycle exists
	node1 := &ListNode{Val: 3}
	node2 := &ListNode{Val: 2}
	node3 := &ListNode{Val: 0}
	node4 := &ListNode{Val: -4}
	node1.Next = node2
	node2.Next = node3
	node3.Next = node4
	node4.Next = node2 // cycle

	if hasCycle(node1) != true {
		fmt.Println("FAIL Test 1")
	} else {
		fmt.Println("PASS Test 1")
	}

	// Test 2: no cycle
	a := &ListNode{Val: 1}
	b := &ListNode{Val: 2}
	a.Next = b

	if hasCycle(a) != false {
		fmt.Println("FAIL Test 2")
	} else {
		fmt.Println("PASS Test 2")
	}

	// Test 3: nil
	if hasCycle(nil) != false {
		fmt.Println("FAIL Test 3")
	} else {
		fmt.Println("PASS Test 3")
	}
}

File: Languages/cpp/algorithms-dsa/leetcode/linked-list/linked_list_cycle.cpp
/*
 * LeetCode 141 - Linked List Cycle
 * Topic: Linked List
 * Difficulty: Easy
 *
 * Floyd's cycle detection: fast pointer moves 2 steps, slow moves 1.
 * If they meet, there's a cycle.
 * Time: O(n)
 * Space: O(1)
 */
#include <iostream>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

bool hasCycle(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return true;
    }
    return false;
}

int main() {
    // Cycle: 3 -> 2 -> 0 -> -4 -> 2 (back to node at index 1)
    ListNode* n1 = new ListNode(3);
    ListNode* n2 = new ListNode(2);
    ListNode* n3 = new ListNode(0);
    ListNode* n4 = new ListNode(-4);
    n1->next = n2; n2->next = n3; n3->next = n4; n4->next = n2;
    assert(hasCycle(n1) == true);

    // No cycle: 1 -> 2
    ListNode* m1 = new ListNode(1);
    ListNode* m2 = new ListNode(2);
    m1->next = m2;
    assert(hasCycle(m1) == false);

    // Single node, no cycle
    assert(hasCycle(new ListNode(1)) == false);

    cout << "All tests passed!" << endl;
    return 0;
}
